### ISSUE TITLE
Initializing builder and solver in Initialize of strategy

### ISSUE THREAD

**Issue by adityaghantasala** [KratosMultiphysics/Kratos#581]
Hello all 

This is a long question. 

I am going through 

> residual_based_newton_raphson_strategy.h

  I noticed that builder and solver is not initialized in the Initialize() function of the strategy. 

All the initializing of the builder and solver is done in InitializeSolutionStep() function. Like the functions 

- SetUpDofSet

-  SetUpSystem

- ResizeAndInitializeVectors 

but the InitializeSolutionStep() of the scheme is called here. Is there a reason why these are ordered in this way ?? 

I am a bit confused here ... aren't SetUpDofSet , SetUpSystem  something for initialization of the builder and solver (and to be re-performed if requested )?? 

I am asking this for the following reason : 

**Previously:**
- I need to work with equationIds  for applying constraints. These were not formulated until SetUpSystem of builder and solver is called. Previously I postponed the formulation of constraint data till I have this information by initially storing VariableKeys and using them later after SetUpSystem is called. 

**Now**
I enhanced the complete constraint application process and incorporated mapper application to obtain master slave relations. Also the constraints can be done through a normal process. Since here it is not good (works only in a serial version) to transfer the Variable or the keys, I am only transferring the master equation ID and corresponding weight. Since the equationIds are not available when the process calls 

> ExecuteInitializeSolutionStep 

it is not possible for me to formulate the constraint data. 

Considering that I am trying to use this implementation MPC in different applications like fluid for periodic boundary conditions, and structures for joints etc ... I think a general solution is necessary.

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee

**Comment by pooyan-dadvand** [comment]
Legacy against us! This is definitely a good question for @RiccardoRossi

**Comment by RiccardoRossi** [comment]
Hi guys,

i am not really sure if this is "legacy". The point is that one typically initializes the strategy (and in there Elements and Conditions) once by calling the function Initialize. The BuilderAndSolver needs to be initialized _at least once_ but possibly multiple times. 

this is currently controlled by the check

       if (pBuilderAndSolver->GetDofSetIsInitializedFlag() == false ||
                mReformDofSetAtEachStep == true)

which is located in the InitializeSolutionStep call.

The scenario for example is that if someone wants to destroy the matrices (and leter reform them) it can call the function "Clear" of the strategy, which will deinitialize the builder and solver.
Elements however should not be reinitialized (for example one might select a reference configuration as the one at which the element initialize is called )

from the practical point of view however we could do

Initialize
InitializeSolutionStep
InitializeBuilderAndSolver <--- this would be new...
...

where the InitializeBuilderAndSolver may not do anything depend on the settings and on wheter it was already called in the past.

would that be correct?
any better solution?

cheers
Riccardo

**Comment by pooyan-dadvand** [comment]
Maybe I'm wrong but is what @adityaghantasala is pointing out is the fact that Initialize of builder and solver is done in InitializeSolutionStep of Strategy which is misleading. I completely agree with @RiccardoRossi that the check in InitializeSolutionStep is necessary but maybe then we should change the name of the Initialize method for more coherence. (Which I'm not sure about its side effects)

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand to my understanding this is not the case. Explaining myself better, the problem is that BuilderAndSolver must be initialized every time it is cleared. Not so for scheme.

i am not telling that the naming is correct or clear... just describing the current behaviour.

**Comment by adityaghantasala** [comment]
I meant what @pooyan-dadvand  said above. 

On the other hand  I understand what @RiccardoRossi  is saying. I think we can also achieve this by doing the same check 

```
   if (pBuilderAndSolver->GetDofSetIsInitializedFlag() == false ||
            mReformDofSetAtEachStep == true)
```

in Initialize() function of the strategy and removing it in the InitializeSolutionStep(). So the builder_and_solver is also initialized (if necessary multiple times) in the initialize function.  In this case a flag is necessary for not doing this "initialization of builder and solver" again for the very first solution step. I tried it in a local branch and this works. 

I am saying this because, in the NewtonRaphson  trategy's InitializeSolutionStep() function, the check 

` if (mSolutionStepIsInitialized == false)`

is not controlling the initializing or reinitializing the builder_and_solver and it also not needed. So we might as well put this 

```
   if (pBuilderAndSolver->GetDofSetIsInitializedFlag() == false ||
            mReformDofSetAtEachStep == true)
```

and following code in the initialize function which makes it more consistent as @pooyan-dadvand  says and also will have the functionality as @RiccardoRossi  says. 

Saying this, I think another function like InitializeBuilderAndSolver is also good and makes things more readable but then I would say it should be called in the Initialize() function of strategy also. 


One more thing I observed is that, in the python fluid solvers "navier_stokes_solver_*.py"  strategy is not initialized in the function Initialize() function. Which I think is also not consistent.

**Comment by RiccardoRossi** [comment]
Hi @adityaghantasala ,

the problem is that "Initialize" of the strategy should be called **strictly once**

**Comment by adityaghantasala** [comment]
@RiccardoRossi  yes. This can also be achieved. I see that now Initialize() is called in every time step when the Solve() is called. But in there there is a check 

`if (mInitializeWasPerformed == false)` 

so we can put this initialization of the builder and solver out side of this if condition and then  the initialization of scheme is done only one and builder and solver is done if necessary.

**Comment by RiccardoRossi** [comment]
@adityaghantasala, sorry i can not follow you.

can u write a pseudo code of what you would like to do?

My biggest problem here is that everyone uses this function, so we need to be very careful not to break existing code.

**Comment by adityaghantasala** [comment]
My bad I was vague in my explanation  ...  here it goes 

**Currently :**

```
virtual void Initialize() override
    {
        KRATOS_TRY;

        if (mInitializeWasPerformed == false)
        {
               ............... ............... ..............
        }
        mInitializeWasPerformed = true;
        
    }

   virtual void InitializeSolutionStep() override
    {
        KRATOS_TRY;

        ........... ............... ................
        if (pBuilderAndSolver->GetDofSetIsInitializedFlag() == false ||
                mReformDofSetAtEachStep == true)
        {
               pBuilderAndSolver->SetUpDofSet(pScheme, BaseType::GetModelPart());
               pBuilderAndSolver->SetUpSystem(BaseType::GetModelPart());
        }

        if (mSolutionStepIsInitialized == false)
        {
                   .................. .............. ................
        }
   }

   virtual double Solve() override
    {
        Initialize(); // ***
        InitializeSolutionStep(); // **
        Predict();
        SolveSolutionStep();
        FinalizeSolutionStep();
        return 0.00;
    }
```
*** Here once initialization of strategy is done in the first call to Solve() .. exactly once as to be done (coz `mInitializeWasPerformed` is set `true`). 

** Here in InitializeSolutionStep() builder and solver is reinitialized if requested.         

**What I am proposing :** 

```
virtual void Initialize() override
    {
        KRATOS_TRY;

        if (mInitializeWasPerformed == false)
        {
               ............... ............... ..............
        }
        mInitializeWasPerformed = true;

       if (pBuilderAndSolver->GetDofSetIsInitializedFlag() == false)
        {   //  initializing the builder and solver
                  pBuilderAndSolver->SetUpDofSet(pScheme, BaseType::GetModelPart());
                  pBuilderAndSolver->SetUpSystem(BaseType::GetModelPart());
        }
    }

   virtual void InitializeSolutionStep() override
    {
        KRATOS_TRY;

        ........... ............... ................
        if (mReformDofSetAtEachStep == true)  
        {  //  reinitializing the builder and solver if requested for
                  pBuilderAndSolver->SetUpDofSet(pScheme, BaseType::GetModelPart());
                  pBuilderAndSolver->SetUpSystem(BaseType::GetModelPart());
        }

        if (mSolutionStepIsInitialized == false)
        {
                   .................. .............. ................
        }
   }

   virtual double Solve() override
    {
        Initialize(); 
        InitializeSolutionStep();
        Predict();
        SolveSolutionStep();
        FinalizeSolutionStep();
        return 0.00;
    }
```

**Comment by adityaghantasala** [comment]
What I was mentioning about the python solvers is 

in the `Initialize() ` function of  for example navier_stokes_solver_vmsmonolithic.py 

```
def Initialize(self):
      ............... 
      ............... 
      # but NO (self.solver).Initialize() // **
```
** To be consistent I think strategy (self.solver) has to be initialized here in all the python solver's `Initialize `function.

**Comment by philbucher** [comment]
Guys is this solved?
I think this might be related to #841 and #1698

**Comment by adityaghantasala** [comment]
Yes this is solved. I think necessary changes were made in C++ strategy.
Secondly it is not directly related to #841 and #1698 .

**Comment by philbucher** [comment]
So can it be closed?

**Comment by adityaghantasala** [comment]
For me yes. !
